include/ginkgo/core/multigrid/multigrid_level.hpp

```diff
@@ -0,0 +1,148 @@
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_MULTIGRID_MULTIGRID_LEVEL_HPP_
+#define GKO_PUBLIC_CORE_MULTIGRID_MULTIGRID_LEVEL_HPP_
+
+
+#include <functional>
+#include <memory>
+
+
+#include <ginkgo/core/base/abstract_factory.hpp>
+#include <ginkgo/core/base/composition.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+namespace gko {
+namespace multigrid {
+
+
+/**
+ * This class represents two levels in a multigrid hierarchy.
+ *
+ * A class implementing this interface should use the
+ * this->get_compositions()->apply(...) as its own apply, which represents
+ * op(b) = prolong(coarse(restrict(b))). All MultigridLevel generator should
+ * inherit from this class.
+ *
+ * @ingroup Multigrid
+ */
+template <typename ValueType>
+class MultigridLevel : public UseComposition<ValueType> {
+public:
+    using value_type = ValueType;
+
+    /**
+     * Returns the operator on fine level.
+     *
+     * @return  the operator on fine level.
+     */
+    std::shared_ptr<const LinOp> get_fine_op() const { return fine_op_; }
+
+    /**
+     * Returns the restrict operator.
+     *
+     * @return  the restrict operator.
+     */
+    std::shared_ptr<const LinOp> get_restrict_op() const
+    {
+        return this->get_operator_at(2);
+    }
+
+    /**
+     * Returns the operator on coarse level.
+     *
+     * @return  the operator on coarse level.
+     */
+    std::shared_ptr<const LinOp> get_coarse_op() const
+    {
+        return this->get_operator_at(1);
+    }
+
+    /**
+     * Returns the prolong operator.
+     *
+     * @return  the prolong operator.
+     */
+    std::shared_ptr<const LinOp> get_prolong_op() const
+    {
+        return this->get_operator_at(0);
+    }
+
+protected:
+    /**
+     * Sets the multigrid level information.
```

Comment:
Minor: maybe add a comment stating that the parameters are stored in that order in the composition.